runtime/reconcilers/alert.go

```diff
@@ -0,0 +1,847 @@
+		if err != nil {
+			return runtime.ReconcileResult{Err: fmt.Errorf("failed to clear trigger: %w", err)}
+		}
+	}
+
+	// Update refs hash
+	if refsTrigger {
+		a.State.RefsHash = refsHash
+		err = r.C.UpdateState(ctx, self.Meta.Name, self)
+		if err != nil {
+			return runtime.ReconcileResult{Err: err}
+		}
+	}
+
+	// Done
+	if a.State.NextRunOn != nil {
+		return runtime.ReconcileResult{Err: executeErr, Retrigger: a.State.NextRunOn.AsTime()}
+	}
+	return runtime.ReconcileResult{Err: executeErr}
+}
+
+// executionSpecHash computes a hash of the alert properties that impact execution.
+// NOTE: Unlike other resources, we don't include the refs' state version in the hash since it's managed separately using refsStateHash.
+func (r *AlertReconciler) executionSpecHash(ctx context.Context, spec *runtimev1.AlertSpec, refs []*runtimev1.ResourceName) (string, error) {
+	hash := md5.New()
+
+	for _, ref := range refs { // Refs are always sorted
+		// Write name
+		_, err := hash.Write([]byte(ref.Kind))
+		if err != nil {
+			return "", err
+		}
+		_, err = hash.Write([]byte(ref.Name))
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if spec.RefreshSchedule != nil {
+		_, err := hash.Write([]byte(spec.RefreshSchedule.TimeZone))
+		if err != nil {
+			return "", err
+		}
+	}
+
+	err := binary.Write(hash, binary.BigEndian, spec.WatermarkInherit)
+	if err != nil {
+		return "", err
+	}
+
+	_, err = hash.Write([]byte(spec.IntervalsIsoDuration))
+	if err != nil {
+		return "", err
+	}
+
+	err = binary.Write(hash, binary.BigEndian, spec.IntervalsCheckUnclosed)
+	if err != nil {
+		return "", err
+	}
+
+	_, err = hash.Write([]byte(spec.QueryName))
+	if err != nil {
+		return "", err
+	}
+
+	_, err = hash.Write([]byte(spec.QueryArgsJson))
+	if err != nil {
+		return "", err
+	}
+
+	_, err = hash.Write([]byte(spec.GetQueryForUserId()))
+	if err != nil {
+		return "", err
+	}
+
+	_, err = hash.Write([]byte(spec.GetQueryForUserEmail()))
+	if err != nil {
+		return "", err
+	}
+
+	if spec.GetQueryForAttributes() != nil {
+		v := structpb.NewStructValue(spec.GetQueryForAttributes())
+		err = pbutil.WriteHash(v, hash)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	err = binary.Write(hash, binary.BigEndian, spec.TimeoutSeconds)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+// refsStateHash computes a hash of the refs and their state versions.
+func (r *AlertReconciler) refsStateHash(ctx context.Context, refs []*runtimev1.ResourceName) (string, error) {
+	hash := md5.New()
+
+	for _, ref := range refs { // Refs are always sorted
+		// Write name
+		_, err := hash.Write([]byte(ref.Kind))
+		if err != nil {
+			return "", err
+		}
+		_, err = hash.Write([]byte(ref.Name))
+		if err != nil {
+			return "", err
+		}
+
+		// Note: Only writing the state version to the hash, not spec version, because it doesn't matter whether the spec/meta changes, only whether the state changes.
+		r, err := r.C.Get(ctx, ref, false)
```

Comment:
Would this cause an issue if the ref was created for the 1st time, went into error state and was fixed?

I remember this case when state version was reset and was the same if it came out of error state vs being created for the 1st time.

Reply:
Good catch! Fixed it